ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala

```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
   private def blind: Step =
     (commitContext, transactionEntry) => {
       val blindingInfo = Blinding.blind(transactionEntry.transaction)
-      buildFinalResult(commitContext, transactionEntry, blindingInfo)
+      buildFinalResultState(commitContext, transactionEntry, blindingInfo)
     }
 
+  private def removeUnnecessaryNodes: Step = (commitContext, transactionEntry) => {
+    val transaction = transactionEntry.submission.getTransaction
+    val nodes: Map[String, TransactionOuterClass.Node] =
+      transaction.getNodesList.asScala.map(n => n.getNodeId -> n)(scala.collection.breakOut)
+    val nodesToKeep: Set[String] = nodes.collect {
+      case (id, node) if node.hasCreate || node.hasExercise => id
+    }(scala.collection.breakOut)
+
+    val roots = transaction.getRootsList.asScala.filter(nodesToKeep)
+
+    val stripUnnecessaryChildren: TransactionOuterClass.Node => TransactionOuterClass.Node = {
```

Comment:
If you make a `def` of this I'm relatively sure the Scala compiler will turn it into a private method of the class. I'm not sure of what happens to function objects. Would this be compiled to a method or would this cause the function object to be created on every call of `removeUnnecessaryNodes`.

Reply:
Looks like on this toy experiment

```scala
object Foo {

  def bar(ns: List[Int]): List[Int] = {
    val predicate: Int => Boolean = _ < 5
    ns.filter(predicate)
  }

}
```

the compiler actually compiles the function to a private method (`$anonfun$bar$1`)

```
> javap -c Foo
Compiled from "Foo.scala"
public final class Foo {
  public static scala.collection.immutable.List<java.lang.Object> bar(scala.collection.immutable.List<java.lang.Object>);
    Code:
       0: getstatic     #18                 // Field Foo$.MODULE$:LFoo$;
       3: aload_0
       4: invokevirtual #20                 // Method Foo$.bar:(Lscala/collection/immutable/List;)Lscala/collection/immutable/List;
       7: areturn
}
> javap -c Foo$
Compiled from "Foo.scala"
public final class Foo$ {
  public static final Foo$ MODULE$;

  public static {};
    Code:
       0: new           #2                  // class Foo$
       3: dup
       4: invokespecial #17                 // Method "<init>":()V
       7: putstatic     #19                 // Field MODULE$:LFoo$;
      10: return

  public scala.collection.immutable.List<java.lang.Object> bar(scala.collection.immutable.List<java.lang.Object>);
    Code:
       0: invokedynamic #41,  0             // InvokeDynamic #0:apply$mcZI$sp:()Lscala/runtime/java8/JFunction1$mcZI$sp;
       5: astore_2
       6: aload_1
       7: aload_2
       8: invokevirtual #47                 // Method scala/collection/immutable/List.filter:(Lscala/Function1;)Lscala/collection/immutable/List;
      11: areturn

  public static final boolean $anonfun$bar$1(int);
    Code:
       0: iload_0
       1: iconst_5
       2: if_icmpge     9
       5: iconst_1
       6: goto          10
       9: iconst_0
      10: ireturn
}
```

Smart compiler. 🙂